### PR TITLE
move Log2, Log10, Ln to datafusion-functions

### LINKED ISSUE #9868: move the Log2, Log10, Ln functions to datafusion-functions crate
### Is your feature request related to a problem or challenge?

As part of #9285, move the Log2, Log10, Ln functions to datafusion-functions crate

### Describe the solution you'd like

Functions are migrated to the new crate and use the new patterns as described in #9286

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by tinfoil-knight**
## Which issue does this PR close?

Closes #9868.

## Rationale for this change

As part of #9285, `log10`, `log2` & `ln` functions should be migrated to the datafusion-functions crate.

Sidenote:
Didn't migrate `log` because some parts (eg: simplifying expressions: see `simpl_log`, `simpl_power`) of `log` & `power` functions are dependent on each other & IMO these should be migrated together.

## What changes are included in this PR?

Extend the `make_math_unary_udf` to accept a parameter for monotonicity & migrate log10, log2, ln from built-in functions to the `datafusion-functions` crate.

## Are these changes tested?

Covered by existing tests.

## Are there any user-facing changes?

No.

### CODE REVIEW

File: datafusion/functions/src/math/mod.rs [hunk 1]
```diff
@@ -24,10 +24,14 @@ mod nans;
 make_udf_function!(nans::IsNanFunc, ISNAN, isnan);
 make_udf_function!(abs::AbsFunc, ABS, abs);
 
-make_math_unary_udf!(TanhFunc, TANH, tanh, tanh);
-make_math_unary_udf!(AcosFunc, ACOS, acos, acos);
-make_math_unary_udf!(AsinFunc, ASIN, asin, asin);
-make_math_unary_udf!(TanFunc, TAN, tan, tan);
+make_math_unary_udf!(Log2Func, LOG2, log2, log2, Some(vec![Some(true)]));
```
Comment (alamb) [inline datafusion/functions/src/math/mod.rs:27]: I wonder if we could make a constant or something that would make this easier to read

For example
```rust
/// Single monotonic argument
const MONOTONIC_ONE: Option<Vec<bool>> = Some(vec![Some(true)])
const NON_MONOTONIC: Option<Vec<bool>> = None;
```

And then 

```suggestion
make_math_unary_udf!(Log2Func, LOG2, log2, log2, MONOTONIC_ONE);
```

Perhaps as a follow on PR

**Review by alamb — APPROVED** [review]
Thank you so much @tinfoil-knight  -- 🚀 

> Didn't migrate log because some parts (eg: simplifying expressions: see simpl_log, simpl_power) of log & power functions are dependent on each other & IMO these should be migrated together.

I agree -- thank you, this makes sense

Note I think we can implement `simpl_log` using `ScalarUDFImpl::simplify` (for example https://github.com/apache/arrow-datafusion/blob/9d47dcaab85294f517c899464c442f9a4518b6f2/datafusion/functions/src/datetime/current_date.rs#L76-L91)

It might be a fun little PR if you wanted to try

File: datafusion/functions/src/math/mod.rs (hunk 1 above)
Comment (tinfoil-knight) [reply datafusion/functions/src/math/mod.rs:27]: @alamb 
I tried this readability fix but Rust doesn't allow allocation (which `vec![]` needs) in `const` or `static` items. Global `let` is also not allowed.

The compiler suggested this with static:
```
consider wrapping this expression in `Lazy::new(|| ...)` from the `once_cell` crate: https://crates.io/crates/once_cell
```

We use `once_cell` in the `physical_plan` crate but introducing it to datafusion functions just for this doesn't seem right to me.

**Comment by tinfoil-knight** [comment]
> Note I think we can implement `simpl_log` using `ScalarUDFImpl::simplify`
 
 Yeah. I'll try this out.

File: datafusion/functions/src/math/mod.rs (hunk 1 above)
Comment (alamb) [reply datafusion/functions/src/math/mod.rs:27]: > We use `once_cell` in the `physical_plan` crate but introducing it to datafusion functions just for this doesn't seem right to me.

That makes sense to me. We would probably have to throw in a `clone()` or something

I thought about this and wrote up another API that would be much nicer https://github.com/apache/arrow-datafusion/issues/9879  (both for this callsite and montonicity in general)

**Pull request merged**